wezterm-gui/src/termwindow/resize.rs

```diff
@@ -302,9 +294,37 @@ impl super::TermWindow {
                 // pixel geometry which is considered to be a user-driven resize.
                 // Stashing the dimensions here avoids that misconception.
                 self.dimensions = dims;
+                // Our new size might not be divisible by the current
+                // increment, and our current size might not be divisible by
+                // the new increment, so we disable the increment while
+                // resizing and then set the new increment afterward.
+                window.set_resize_increments(1, 1, 0, 0);
```

Comment:
one thing I'm not 100% sure about here is the ordering of this wrt. to the final (re)application of the configured values below.  What I mean by that is that this conditional block ("Queue up...") has effects that happen partially asynchronously before they fully apply and settle, triggering this resize logic again in a later pass.  For example, tiling window managers can and do re-adjust the size, and on macOS, if the window would exceed the screen size, it will change it.

I don't know if that matters here: I think resize increments are not really compatible with tiling window managers, so I don't know if we need to sweat the details of what happens there and just tell folks to pick which paradigm they want to use.

Reply:
I think that there are at least two invariants we want to be maintained.  First is we want the window manager to always see a consistent version of (1) our requested resize increments and (2) our requested window size.  By consistent here, I mean that our resize increments should divide the window size.  I *think* we are OK with this invariant, at least if we make the basic assumption that the window manager is processing events in the order delivered (if it's not doing that then I don't know how we can ever guarantee anything).

But I think that there is a second one that involves consistency between the client's (wezterm's) internal state and ConfigureNotify events, especially if those ConfigureNotify events come in "late".  For instance, if I press ctrl+-, normally something like this would happen (take scale to be a scaling factor, termsize to be terminal dimensions as measured in cells, and windowsize to be the window dimensions as measured in pixels):

* Initial (scale = scale0, termsize = termsize0, windowsize = windowsize0)
* Ctrl+- (scale = scale1, send ConfigureRequest(scale1 * termsize))
* ConfigureNotify (windowsize = scale1 * termsize, termsize = windowsize1 / scale1)

So far so good!

What if I press ctrl+- twice very quickly?

* Initial (scale = scale0, termsize = termsize0, windowsize = windowsize0)
* Ctrl+- (scale = scale1, send ConfigureRequest(scale1 * termsize))
* Ctrl+- (scale = scale2, send ConfigureRequest(scale2 * termsize))
* ConfigureNotify (windowsize = scale1 * termsize, termsize = windowsize1 / scale2) // WRONG termsize
* ConfigureNotify (windowsize = scale2 * termsize, termsize = windowsize2 / scale2) // correct termsize is restored

OK, we have the wrong terminal size for a moment, but at least it is eventually corrected.  But what if I'm spamming the key and this happens:

* Initial (scale = scale0, termsize = termsize0, windowsize = windowsize0)
* Ctrl+- (scale = scale1, send ConfigureRequest(scale1 * termsize))
* Ctrl+- (scale = scale2, send ConfigureRequest(scale2 * termsize))
* ConfigureNotify (windowsize = scale1 * termsize, termsize = windowsize1 / scale2) // WRONG termsize
* Ctrl+- (scale = scale3, send ConfigureRequest(scale3 * termsize_wrong)) // ConfigureRequest sent with WRONG size
* ConfigureNotify (windowsize = scale3 * termsize, termsize = windowsize2 / scale3) // a different wrong termsize
* ConfigureNotify (windowsize = scale3 * termsize_wrong, termsize = windowsize3 / scale3 = termsize_wrong) // original wrong termsize :(

So I believe that we also want some invariant which is something like that we have consistent scale and window sizes whenever we send ConfigureRequests, which means that we will need to somehow be sure that we have received all pending ConfigureNotify requests from the window manager.

And actually, this race has nothing to do with this merge request.  I can reproduce this race condition, or at least one like it, on the main branch without even using a tiling window manager, just by spamming the ctrl+- and ctrl+= keys.  Specifically, spamming ctrl+- or ctrl+= will eventually change my termsize.

If this race condition exists, then this might be tricky to fix, as I'm not sure what synchronization primitive to use to fix this.  If `XSync()` is under the hood sending an `GetInputFocus` request and waiting for the response, then what would `XSyncWindowManager()` look like?  Create an unmapped dummy window, reconfigure it, and then wait for the `ConfigureNotify` event?  If we make the same basic assumption as above that the window manager is processing events in the order delivered, then it might work...

Reply:
FWIW, we're using xcb rather than libx11 under the hood, so there is no XSync.  Also: this synchronization problem potentially applies to other platforms where the model is different. For example, on Windows, we may receive multiple messages one after the other, or we may get recursively called by windows' internal message dispatching when we trigger a size change while processing an event.

So... not easy to resolve in isolation.

In the context of this PR: if your changes work about as well as the existing state of things, then I'm not going to require you to resolve this synchronization issue here.

Reply:
> FWIW, we're using xcb rather than libx11 under the hood, so there is no XSync. 

Of course -- I was just speaking in terms of pseudocode there, although if I recall correctly, XCB also uses GetInputFocus to implement synchronization in the *_checked variants of its functions.

> Also: this synchronization problem potentially applies to other platforms where the model is different. For example, on Windows, we may receive multiple messages one after the other, or we may get recursively called by windows' internal message dispatching when we trigger a size change while processing an event.
> 
> So... not easy to resolve in isolation.

True. But this is really bugging me now.  I'd like to get back to it at some point, if even just for the X11 case.

> In the context of this PR: if your changes work about as well as the existing state of things, then I'm not going to require you to resolve this synchronization issue here.

Acknowledged, and thanks for your other feedback too.  It was very helpful.  I'll implement the recommendations as soon as I get a chance.